File: fetch/rust/src/tides.rs
// https://www.admiralty.co.uk/digital-services/data-solutions/uk-tidal-api

use reqwest::{Client, header};

pub fn bosham_location_id() -> String {
    std::env::var("BoshamId").expect("BoshamId should be available as an environment variable")
}
fn subscription_key() -> String {
    std::env::var("SubscriptionKey").expect("Subscription key should be available as an environment variable")
}

fn admiralty_stations_url() -> String {
    "https://admiraltyapi.azure-api.net/uktidalapi/api/V1/Stations".to_owned()
}

fn admiralty_tidal_events_url(station_id: String, duration: i32) -> String {
    format!("https://admiraltyapi.azure-api.net/uktidalapi/api/V1/Stations/{stationId}/TidalEvents?{duration}", 
        stationId = station_id, 
        duration = duration)
}

fn admiralty_headers() -> header::HeaderMap {
    let mut headers = header::HeaderMap::new();
    let mut auth_value = header::HeaderValue::from_str(&subscription_key()).unwrap();
    auth_value.set_sensitive(true);
    headers.insert("Ocp-Apim-Subscription-Key", auth_value);
    headers
}

pub async fn admiralty_locations() -> reqwest::Result<String> {
    let client = Client::builder().default_headers(admiralty_headers()).build()?;
    let response = client.get(&admiralty_stations_url()).send().await?;
    response.text().await
}

pub async fn admiralty_location(id: String) -> reqwest::Result<String> {
    let client = Client::builder().default_headers(admiralty_headers()).build()?;
    let url = format!("{domain}?{stationId}", domain = &admiralty_stations_url(), stationId = id);
    let response = client.get(url).send().await?;
    response.text().await
}

pub async fn admiralty_tidal_events(id: String, duration: i32) -> reqwest::Result<String> {
    let client = Client::builder().default_headers(admiralty_headers()).build()?;
    let response = client.get(&admiralty_tidal_events_url(id, duration)).send().await?;
    response.text().await
}

File: fetch/src/Tides.cs
using System;
using System.IO;
using System.Net.Http;
using System.Threading.Tasks;

namespace DailyTide
{
    public class Tides
    {
        private readonly string SubscriptionKey = Environment.GetEnvironmentVariable("SubscriptionKey");
        private readonly string StationsUrl = "https://admiraltyapi.azure-api.net/uktidalapi/api/V1/Stations";

        private readonly HttpClient Client;

        public Tides(HttpClient client) 
        {
            Client = client;
        }

        public async Task<string> GetTideEvents(string locationId)
        {
            Client.DefaultRequestHeaders.Add("Ocp-Apim-Subscription-Key", SubscriptionKey);
            var request = $"{StationsUrl}/{locationId}/TidalEvents?7";
            Console.WriteLine($"making request to {request}");
            var response = await Client.GetAsync(request);
            if(!response.IsSuccessStatusCode)
            {
                Console.Error.WriteLine( response.StatusCode );
                Console.Error.WriteLine( await response.Content.ReadAsStringAsync() );
                throw new Exception("error with api request");
            }
            return await response.Content.ReadAsStringAsync();
        }

        public async Task<string> GetLocations()
        {
            Client.DefaultRequestHeaders.Add("Ocp-Apim-Subscription-Key", SubscriptionKey);
            var request = $"{StationsUrl}";
            Console.WriteLine($"making request to {request}");
            var response = await Client.GetAsync(request);
            if(!response.IsSuccessStatusCode)
            {
                Console.Error.WriteLine( response.StatusCode );
                throw new Exception("error with api request");
            }
            return await response.Content.ReadAsStringAsync();
        }
    }
}